Implement the python side of SFTP access to all sharelinks

jonasbardino opened the pull request:
Implement SFTP access to all sharelinks with added server-side access limitations on all operations when in read-only or write-only mode.

~This is still work in progress and the PAM and NSS bits are still missing for sftpsubsys support.~
We may still want to look into pointing all read-only sharelink access to a separate read-only (bind-) mount of the persistent storage file system to be extra protected against any missing access limitations.

As @aputtu pointed out after initial testing of the write-only sharelinks they felt somewhat crippled when used e.g. mounted from a file manager like `Nautilus` or `Thunar` because we initially rejected **all** read operations including  `stat`/`lstat` calls to query basic file and folder info. After some reconsideration that restriction was loosened up to allow explicit `stat`/`lstat` calls on explicitly given paths to make most clients behave. Similarly it turned out that simple get downloads from `sftp` on read-only shares would fail when empty `chattr` calls were rejected. So such non-changing `chattr` calls are now also allowed. 
We may still want to similarly allow non-changing `chown`/`chmod` calls and perhaps even override `list_folder` to always just return an empty list for write-only shares.

jonasbardino committed:
Implement the python side of SFTP access to all sharelinks with server-side
access limitations on all operations when in read-only or write-only mode.

This is work in progress and the PAM and NSS bits are still missing for
sftpsubsys support.

jonasbardino committed:
Extend PAM+NSS sharelink access to cover all three access modes in order to
support access from sshd+sftpsubsys service, too.

Simply adjusts the existing read-write sharelink test and auth to instead loop
through the three access modes in turn and do the similar test for each. Minor
adjustments to continue until one succeeds or all have been tried.

The existing test+auth code is purposefully kept unindented in this version to
make the diff simple to review. Will follow-up with a separate commit which
only indents to fit loop nesting to keep that change simple as well.

jonasbardino committed:
Avoid string overflow for mode strings of size at most 10 chars plus one byte
string termination (`\0`). Dumb down the sharelink mode iterator to keep it
simple.
Adjust prefix for shared PAM and NSS log line helper to clarify that. Otherwise
we get `pam_mig` log entries even for key login, which never hits PAM.

jonasbardino commented:
With PR #286 and #287 merged there are only a few remaining lint errors in `grid_sftp.py` and as mentioned in PR #287 they at least look rather harmless. We should still either fix or whitelist them with `pylint` annotations.

jonasbardino committed:
Elaborate the grid_sftp sharelink access handling for a better and more
consistent client experience.
Basically filter certain operations like `chattr` and `stat`/`lstat` more
carefully in order to let read-only and write-only shares behave.

For read-only shares it is necessary to allow non-changing `chattr` calls to
avoid `sftp` failing in simple `get path` downloads.
For write-only shares it's necessary to allow `lstat`/`stat` on given files and
folders to avoid errors e.g. when sftp or sshfs mount `rm`s a file.

Additional tweaks to allow non-changing `chmod`/`chown` in read-only shares are
still under consideration.

rasmunk commented on mig/shared/griddaemons/login.py:
```diff
@@ -682,7 +694,7 @@ def refresh_job_creds(configuration, protocol, username):
 
 def refresh_share_creds(configuration, protocol, username,
-                        share_modes=['read-write']):
+                        share_modes=[READ_WRITE_ACCESS]):
```
I'm not a big fan of the mutable object share_modes=[] as a default argument. It's not an issue in this specific instance but it can cause issues later if any modification applied to share_modes inside of refresh_share_creds will persist across calls https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

jonasbardino replied:
Good point, thanks. Changed to tuple in latest rev as I would prefer to keep the default clearly visible.

jonasbardino replied:
I've pushed another round of similar adjustments to avoid the mutable default args as we talked about off-list.

rasmunk commented on mig/shared/griddaemons/login.py:
```diff
@@ -694,6 +706,7 @@ def refresh_share_creds(configuration, protocol, username,
     have guards in place to support read-only or write-only mode in daemons.
```
Just a polishing thing. The NOTE about only supporting read-write sharelinks would have to be updated before merging due to the changes in this PR

jonasbardino replied:
Updated in latest rev.

jonasbardino committed:
Specifically use stdint types instead of uint in a few locations as suggested
by @rasmunk.

jonasbardino committed:
Address PR feedback - thanks @rasmunk .

jonasbardino committed:
Adjust the calling side share_modes to use tuple, too.
Minor layout adjustments from editor/autopep8.

jonasbardino committed:
Fix another case of C99 variable declaration like PR310.

jonasbardino committed:
Sync latest minor fix in NSS to PAM.

jonasbardino committed:
Rework the use of the `args` parameter to avoid mutable default arguments as
suggested by @rasmunk . Also simplifies the calls a bit where no args are used.

Martin-Rehr reviewed (approved):
Approved

jonasbardino commented:
I've added the consideration bits from the PR description to the [milestone](https://github.com/ucphhpc/migrid-sync/milestone/9) and will proceed with the merge.
